### PR TITLE
KLDiv for Model

### PR DESCRIPTION
**Pull request by YanickZengaffinen**
Can now specify if the leader should be trained with L2-loss over all individual predictions or with KLDiv loss over all samples

See https://github.com/nilscrm/stackelberg-ml/issues/27

### CODE REVIEW

File: stackelberg_mbrl/envs/querying_env.py [hunk 1]
```diff
@@ -106,23 +118,53 @@ def step(self, action: np.ndarray):
...
+
+                # If this is the last step give the model reward based on the average prediction loss
+                # We do an average instead of a sum so that the model is not encouraged to play short games.
+                if terminated or truncated:
+                    reward = -self.total_loss / (self.step_count - len(self.queries) + 1)
+            elif self.reward == "kldiv":            
+                s = self.current_state
+                a = self.policy_action
+
+                self.true_P_traj_total[s, a] += one_hot(true_env_obs, self.true_env_num_states)
```
Comment (nilscrm) [inline stackelberg_mbrl/envs/querying_env.py:143]: In our case, especially for the the first state which we only visit once, KL div does not really make a difference right? Cause we still compare it to only one observation.

File: stackelberg_mbrl/envs/querying_env.py [hunk 2]
```diff
@@ -106,23 +118,53 @@ def step(self, action: np.ndarray):
...
+                        true_total = torch.tensor(self.true_P_traj_total)
+                        model_total = torch.tensor(self.model_P_traj_total)
+                        obs_cnt = torch.tensor(self.obs_P_cnt)
+
+                        # convert to estimates P(-|s,a) forall s,a
+                        true_P = true_total / obs_cnt.unsqueeze(-1)
+                        model_P = model_total / obs_cnt.unsqueeze(-1)
+
+                        # compute D_KL(P(-|s,a), \hat{P}(-|s,a)) forall s,a
+                        kl_divs = (true_P * (true_P / (model_P + 1e-5)).log()).sum(-1)
```
Comment (nilscrm) [inline stackelberg_mbrl/envs/querying_env.py:158]: I think this is not quite the correct expression because this estimate basically weights every state/action pair equally. So according to the paper (and what we've been doing so far) we need to sample from mu but here we're basically sampling uniformely.

Comment (nilscrm) [inline stackelberg_mbrl/envs/querying_env.py:158]: The zeros in the log also feel kinda illegal xD.

**Review by nilscrm — COMMENTED** [review]
Thinking about what this approach leads to, it seems to me as if we're basically creating another world model (in our case with `true_total`). This will clearly only work for very low dimensional settings where we can sample all state/action pairs we visits often. And otherwise this will lead to creating another actual world model which would probably defeat the purpose of the first one right?

File: stackelberg_mbrl/experiments/experiment_config.py [hunk 3]
```diff
@@ -19,10 +19,12 @@ class LoadPolicy(BaseModel):
     path: FilePath
 
 class WorldModelConfig(BaseModel):
+    leader_reward: str = "kldiv"
     total_training_steps: int = 1_000_000
     model_save_name: str | None = None
 
 class LoadWorldModel(BaseModel):
+    leader_reward: str = "kldiv"
```
Comment (nilscrm) [inline stackelberg_mbrl/experiments/experiment_config.py:27]: Should probably be a config for the `LeaderEnv` just like [here](https://github.com/nilscrm/stackelberg-ml/blob/5be143a62afd7fe20937df135e0f7d98f67dbb07/stackelberg_mbrl/experiments/experiment_config.py#L22.)

**Review by nilscrm — COMMENTED** [review]
The `LeaderEnv` is getting quite big with reward weighting and different divergences. Do you think it makes sense to have just two kinds of functions that the env calls, one that that calculates rewards per step and the other calculating reward at the end given the entire trajectory?
